modules/services/window-managers/bspwm/default.nix

```diff
@@ -0,0 +1,77 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.xsession.windowManager.bspwm;
+  bspwm = cfg.package;
+
+  camelToSnake = s:
+    builtins.replaceStrings lib.upperChars (map (c: "_${c}") lib.lowerChars) s;
+
+  formatConfig = n: v:
+    let
+      formatList = x:
+        if isList x
+        then throw "can not convert 2-dimensional lists to bspwm format"
+        else formatValue x;
+
+      formatValue = v:
+        if isBool v then (if v then "true" else "false")
+        else if isList v then concatMapStringsSep ", " formatList v
+        else if isString v then "'${lib.strings.escapeShellArg v}'"
+        else toString v;
+    in
+      "bspc config ${n} ${formatValue v}";
+
+  formatMonitors = n: v: "bspc monitor ${n} -d ${concatStringsSep " " v}";
+
+  formatRules = target: directiveOptions:
+    let
+      formatDirective = n: v:
+        if isBool v then (if v then "${camelToSnake n}=on" else "${camelToSnake n}=off")
+        else if (n == "desktop" || n == "node") then "${camelToSnake n}='${v}'"
+        else "${camelToSnake n}=${lib.strings.escapeShellArg v}";
+
+      directives = filterAttrs (n: v: v != null && !(lib.strings.hasPrefix "_" n)) directiveOptions;
+      directivesStr = builtins.concatStringsSep " " (mapAttrsToList formatDirective directives);
+    in
+      "bspc rule -a ${target} ${directivesStr}";
+
+  formatStartupPrograms = map (s: "${s} &");
+
+in
+
+{
+  options = import ./options.nix { inherit pkgs; inherit lib; };
+
+  config = mkIf cfg.enable (mkMerge [
+    {
+      home.packages = [ bspwm ];
+      xsession.windowManager.command = "${cfg.package}/bin/bspwm";
+    }
+
+    (mkIf (cfg.config != null) {
+      xdg.configFile."bspwm/bspwmrc" = {
```

Comment:
Could do
```
xdg.configFile."bspwm/bspwmrc".source =
  pkgs.writeShellScript "bspwmrc"
  (concatStringsSep "\n" (
    mapAttrsToList formatMonitors cfg.monitors
    ++ …
  );
```
instead to make sure the right shell gets used and the syntax of the generated script is checked.

Reply:
Actually, skip this and instead do something like
```
xsession.windowManager.command =
  let
    configCmdOpt = optionalString (cfg.config != null) "-c ${configFile}";
  in
    "${cfg.package}/bin/bspwm ${configCmdOpt}";
```
where
```
configFile =
  pkgs.writeShellScript "bspwmrc"
  (concatStringsSep "\n" (
    mapAttrsToList formatMonitors cfg.monitors
    ++ …
  );
```

Reply:
`writeShellScript` sounds good. will the second suggestion put the config file in the nix store, rather than its xdg position? If that's the case, is that standard practice? I quite like being able to look up config files in their usual positions, even if managed by home-manager.

Reply:
Yes, it would refer to the Nix store path directly rather than implicitly through the symbolic link at the XDG location. Doing it this way improves atomicity and in the choice between atomicity and standard file location we should always prefer atomicity. This is a fundamental idea behind NixOS (and by extension, Home Manager).

To be fair, I think we have missed this in a few places here and there in HM but I would like to fix those.

Reply:
Oof, that's a tough tradeoff. Will take some getting used to :\

Passing generational files as arguments to long-running applications might also be troublesome wrt garbage collection, unless we can be certain they will only be read at startup time. In this case for example, there is `bspc wm --restart` that I believe will read the config file again.

Reply:
Saw that this is what nixpkgs itself does in several places. Still not 100% convinced, but it's the upstream way :) I'll adapt my PR to match